### PR TITLE
[FLINK-5220] [Table API & SQL] Flink SQL projection pushdown

### PR DESCRIPTION
**Pull request by beyond1920**
This pr aims to do projection pushdown optimization.
The main changes including:
1. add  PushProjectIntoBatchTableSourceScanRule to match DataSetCalc->BatchTableSourceScan
2. add ProjectableTableSource to represent a TableSource which supports Projection pushdown
3. change BatchScan cost compute logic
4. add a test case

### CODE REVIEW

**Comment by fhueske** [comment]
Hi @beyond1920 and @KurtYoung,

please see my [comment](https://github.com/apache/flink/pull/2810#issuecomment-265180294) on PR #2810 for how to continue with this PR.
Please let me know what you think.

Best, Fabian

**Comment by beyond1920** [comment]
@fhueske , thanks for your review. I have already modified the pr based on your advice, including following main changes:
1. Adapt the cost model for the BatchTableSourceScan
2. Modify DataSetCalcConverter -> RexProgramProjectExtractor, and add a test case
Thanks.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchScan.scala [hunk 1]
```diff
@@ -45,7 +45,8 @@ abstract class BatchScan(
   override def computeSelfCost (planner: RelOptPlanner, metadata: RelMetadataQuery): RelOptCost = {
 
     val rowCnt = metadata.getRowCount(this)
-    planner.getCostFactory.makeCost(rowCnt, rowCnt, 0)
+    val columnCnt = getRowType.getFieldCount
```
Comment (beyond1920) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchScan.scala]: Yes, that's right. Thanks.

**Comment by tonycox** [comment]
Hi @fhueske, this PR looks good to me

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/PushProjectIntoBatchTableSourceScanRule.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  override def onMatch(call: RelOptRuleCall) {
+    val calc: DataSetCalc = call.rel(0).asInstanceOf[DataSetCalc]
+    val scan: BatchTableSourceScan = call.rel(1).asInstanceOf[BatchTableSourceScan]
+
+    val usedFields: Array[Int] = extractRefInputFields(calc.calcProgram)
+
+    // if no fields can be projected, there is no need to transform subtree
+    scan.tableSource.getNumberOfFields match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/PushProjectIntoBatchTableSourceScanRule.scala]: Can you convert this into a simple `if` condition?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/plan/rules/util/RexProgramProjectExtractorTest.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    val t1 = rexBuilder.makeInputRef(types.get(1), 1)
+    val t2 = rexBuilder.makeInputRef(types.get(3), 3)
+    val t3 = builder.addExpr(rexBuilder.makeCall(SqlStdOperatorTable.MULTIPLY, t0, t2))
+    val t4 = rexBuilder.makeExactLiteral(BigDecimal.valueOf(100L))
+    val t5 = rexBuilder.makeExactLiteral(BigDecimal.valueOf(6L))
+    // project: amount, id, amount * price
+    builder.addProject(t0, "amount")
+    builder.addProject(t1, "id")
+    builder.addProject(t3, "total")
+    // condition: amount * price < 100 and amount > 6
```
Comment (beyond1920) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/plan/rules/util/RexProgramProjectExtractorTest.scala]: good point, thanks

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/ProjectableTableSourceITCase.scala [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      BasicTypeInfo.STRING_TYPE_INFO,
+      BasicTypeInfo.LONG_TYPE_INFO,
+      BasicTypeInfo.INT_TYPE_INFO,
+      BasicTypeInfo.DOUBLE_TYPE_INFO),
+    fieldNames = Array[String]("name", "id", "amount", "price")
+  )
+
+  /** Returns the data of the table as a [[org.apache.flink.api.java.DataSet]]. */
+  override def getDataSet(execEnv: JavaExecEnv): JavaSet[Row] = {
+    execEnv.createInput(new ProjectableInputFormat(33, fieldNames), getReturnType).setParallelism(1)
```
Comment (beyond1920) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/ProjectableTableSourceITCase.scala]: yes, thanks

**Comment by beyond1920** [comment]
@fhueske , thanks for your review, I already modify code based on your advices.

**Comment by fhueske** [comment]
Thanks for the update @beyond1920.
The PR is good to merge :-)

**Comment by fhueske** [comment]
Merging

**Pull request closed without merging**